### ISSUE TITLE
Add timeouts (as a kind "exception")

### ISSUE THREAD

**Issue by jeapostrophe** [AlacrisIO/meta#44]
Rather than implementing timeouts with a more fundamental "choice" operation, instead think about them as a kind of exception that could be thrown and use an effect-style system to ensure that no exceptions reach the top-level. I can't think of other kinds of interrupts now, but perhaps there are things like network splits that make sense in the same framework.

**Comment by fare** [comment]
Yes, especially after my work on state channels, it appears that there must be a special provision for timeouts as exceptions.

In the case of a two-player interaction, it's simple, and there is one obvious exception handler for all two-player interactions: everyone has a "fully owned" amount that is returned to them if they timeout, the rest being the fully owned amount of the other one, and the pot+collaterals together. Logical requirement: sum of fully owned plus pot+collaterals equals resources under contract control.

In the case of more than two participants, it's more complex, and in general requires a developer-provided exception handler, because the timed out participant may have been holding information necessary to continue the interaction, and the remaining participants need to be able to either roll-back or roll-forward this interaction without the cooperation of the timed out participant—and of further future timed out participants!

**Comment by fare** [comment]
For now, I propose we hardcode the two-player solution in the compiler.

Eventually, we ought to be able to express the general case of developer-supplied exception handler, and implement the two-player solution as a standard library function written in the language itself.
